src/RequestParser.cpp

```diff
@@ -0,0 +1,724 @@
+	if (headerLine != "\r") {
+		request.errorCode = StatusBadRequest;
+		throw std::runtime_error(ERR_MISS_CRLF);
+	}
+
+	// Step 3: Parse body (if any)
+	if (m_hasBody) {
+		if (m_chunked)
+			parseChunkedBody(request);
+		else
+			parseNonChunkedBody(request);
+	}
+}
+
+/**
+ * @brief Parses the HTTP method from the request line.
+ *
+ * This function extracts the HTTP method from the beginning of the provided request line.
+ * It supports the "GET", "POST", and "DELETE" methods. The method is stored in `m_request.method`,
+ * and a corresponding Method is set in `m_requestMethod`. If the method is not recognized,
+ * an error code is set to StatusMethodNotImplemented and a `std::runtime_error` is thrown.
+ *
+ * @param requestLine The request line string from which the method is to be parsed.
+ * @param request The HTTP request object to be filled.
+ * @return A substring of `requestLine` starting from the character after the parsed method.
+ * @throws std::runtime_error If the method is not implemented, an error is thrown and the error code is set to
+ * StatusMethodNotImplemented.
+ */
+std::string RequestParser::parseMethod(const std::string& requestLine, HTTPRequest& request)
+{
+	int index = 0;
+
+	if (requestLine.compare(0, getLength, "GET") == 0) {
+		request.method = MethodGet;
+		index = getLength;
+	} else if (requestLine.compare(0, postLength, "POST") == 0) {
+		request.method = MethodPost;
+		index = postLength;
+	} else if (requestLine.compare(0, deleteLength, "DELETE") == 0) {
+		request.method = MethodDelete;
+		index = deleteLength;
+	} else {
+		request.errorCode = StatusMethodNotImplemented;
+		throw std::runtime_error(ERR_METHOD_NOT_IMPLEMENTED);
+	}
+	return (requestLine.substr(index));
+}
+
+/**
+ * @brief Parses the URI from the request line.
+ *
+ * This function extracts the URI from the provided request line.
+ * It validates the URI, ensuring it starts with a slash ('/'),
+ * and checks each character for validity.
+ * If the URI contains a query ('?') or fragment ('#'), it delegates parsing to
+ * `parseUriQuery` and `parseUriFragment` respectively.
+ * If an invalid character is encountered, an error code is set, and
+ * a `std::runtime_error` is thrown.
+ *
+ * @param requestLine The request line string from which the URI is to be parsed.
+ * @param request The HTTP request object to be filled.
+ * @return A substring of `requestLine` starting from the character after the parsed URI.
+ * @throws std::runtime_error If the URI is missing the initial slash,
+ *         contains invalid characters, or any other URI-related errors occur.
+ */
+std::string RequestParser::parseUri(const std::string& requestLine, HTTPRequest& request)
+{
+	int index = 0;
+	if (requestLine.at(index) != '/') {
+		request.errorCode = StatusBadRequest;
+		throw std::runtime_error(ERR_URI_MISS_SLASH);
+	}
+
+	// Check URI string for invalid chars
+	std::string::const_iterator delimiterPos = find(requestLine.begin(), requestLine.end(), ' ');
```

Comment:
std::string::const_iterator delimiterPos can be made const.

Reply:
Changed.